### PR TITLE
Update cdkExecPolicy.yaml to cleanup overly excessive permissions

### PR DESCRIPTION
**Pull request by mourya-33**
### Feature or Bugfix
BugFix

### Detail
This PR is to address overly excessive permissions for cdk custom execution policy for linking accounts that are reported by checkov.

### Relates
https://github.com/data-dot-all/dataall/issues/876

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized? N/A
  - What precautions are you taking before deserializing the data you consume? N/A
  - Is injection prevented by parametrizing queries? N/A
  - Have you ensured no `eval` or similar functions are used? N/A
- Does this PR introduce any functionality or component that requires authorization? N/A
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms? N/A
  - Are you logging failed auth attempts? N/A
- Are you using or adding any cryptographic features? N/A
  - Do you use a standard proven implementations? N/A
  - Are the used keys controlled by the customer? Where are they stored? N/A
- Are you introducing any new policies/roles/users? N/A
  - Have you used the least-privilege principle? How? Yes. Restricted overly permissive permissions to required resources.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: deploy/cdk_exec_policy/cdkExecPolicy.yaml [hunk 1]
```diff
@@ -196,9 +206,7 @@ Resources:
             Effect: Allow
             Action:
               - 'cloudformation:*'
-            Resource:
```
Comment (zsaltys) [inline deploy/cdk_exec_policy/cdkExecPolicy.yaml:199]: Why did you change this?

File: deploy/cdk_exec_policy/cdkExecPolicy.yaml [hunk 2]
```diff
@@ -74,7 +74,17 @@ Resources:
               - 'quicksight:SetGroupMapping'
               - 'quicksight:RegisterUser'
               - 'quicksight:Describe*'
-            Resource: '*'
+            Resource:
```
Comment (zsaltys) [inline deploy/cdk_exec_policy/cdkExecPolicy.yaml:77]: This is going to solve the checkov scan problem but technically this is not safe because it allows access to all quicksight resources on the account. But like you said there's no way to limit by prefix? We don't use quicksight at the moment so it's not a massive issue for us. I also hope when we modularize this file we can simply skip this statement altogether. Maybe the data.all team can comment if we can make this more restrictive?

**Review by zsaltys — CHANGES_REQUESTED** [review]

File: deploy/cdk_exec_policy/cdkExecPolicy.yaml [hunk 3]
```diff
@@ -253,13 +261,16 @@ Resources:
...
               - 'codecommit:Create*'
               - 'codecommit:DeleteRepository'
               - 'codecommit:*Tag*'
               - 'codebuild:*Project*'
-            Resource: '*'
+            Resource: 
+               - !Sub 'arn:${AWS::Partition}:codecommit:${AWS::Region}:${AWS::AccountId}:${EnvironmentResourcePrefix}*'
+               - !Sub 'arn:${AWS::Partition}:codepipeline:${AWS::Region}:${AWS::AccountId}:${EnvironmentResourcePrefix}*'
+               - !Sub 'arn:${AWS::Partition}:codebuild:${AWS::Region}:${AWS::AccountId}:project/${EnvironmentResourcePrefix}*'
+               - !Sub 'arn:${AWS::Partition}:codebuild:${AWS::Region}:${AWS::AccountId}:project/DDKCodePipelineBuildSynthCd*'
```
Comment (noah-paige) [inline deploy/cdk_exec_policy/cdkExecPolicy.yaml:272]: can we just restrict this by `DDK*` prefix

Also what version of data.all are you testing CDK pipeline creation on?

File: deploy/cdk_exec_policy/cdkExecPolicy.yaml (hunk 2 above)
Comment (noah-paige) [reply deploy/cdk_exec_policy/cdkExecPolicy.yaml:77]: I think we can add conditions to only add particular statements if the feature is enabled - i'll work with mourya and see if we can update this PR accordingly

File: deploy/cdk_exec_policy/cdkExecPolicy.yaml (hunk 1 above)
Comment (mourya-33) [reply deploy/cdk_exec_policy/cdkExecPolicy.yaml:199]: @zsaltys , this  was a sync issue on my end, thank you for notifying. I reverted this to previous permission set now.

**Comment by dlpzx** [comment]
Hi @mourya-33! It looks like this PR is getting a bit outdated. Can you merge the latest changes from `main` into the working branch?

**Comment by mourya-33** [comment]
> Hi @mourya-33! It looks like this PR is getting a bit outdated. Can you merge the latest changes from `main` into the working branch?

done @dlpzx

File: frontend/src/modules/Environments/views/EnvironmentCreateForm.js [hunk 4]
```diff
@@ -395,7 +396,7 @@ const EnvironmentCreateForm = (props) => {
...
+                                text={`aws cloudformation --region REGION create-stack --stack-name DataAllCustomCDKExecPolicyStack --template-body file://cdkExecPolicy.yaml --parameters ParameterKey=EnvironmentResourcePrefix,ParameterValue=dataall ParameterKey=NotebooksEnabled,ParameterValue=false ParameterKey=MLStudioEnabled,ParameterValue=false ParameterKey=PipelinesEnabled,ParameterValue=false ParameterKey=DashboardsEnabled,ParameterValue=false --capabilities CAPABILITY_NAMED_IAM && aws cloudformation wait stack-create-complete --stack-name DataAllCustomCDKExecPolicyStack --region REGION && cdk bootstrap --trust ${trustedAccount} -c @aws-cdk/core:newStyleStackSynthesis=true --cloudformation-execution-policies arn:aws:iam::ACCOUNT_ID:policy/DataAllCustomCDKPolicy aws://ACCOUNT_ID/REGION`}
```
Comment (noah-paige) [inline frontend/src/modules/Environments/views/EnvironmentCreateForm.js:399]: I think maybe we leave `--parameters` empty here and in the CloudFormation template default to true 

So will be the same command as before and by default will give all permissions. Then can have documentation that discusses how to restrict further if needed

**Comment by noah-paige** [comment]
Testing CDK Exec Policy

- [ ] CICD Pipeline Successful
- [ ] Download CDK Exec Policy File from UI
- [ ] Deploy CDK Exec Policy (All Params False)
- [ ] Create Env Stack Successfully (All Feature Modules Off)
- ML Studio Test
    - [ ] Update CDK Exec Policy to MLStudioEnabled=true
    - [ ] Permissions Added to CDK Exec Policy
    - [ ] Enable Env ML Studio Feature + Studio Domain Created Successfully
    - [ ] Create ML Studio User Profile Successfully
    - [ ] Delete ML Studio User Profile Successfully
    - [ ]  Disable Env ML Studio Feature Successfully
- Notebook Test 
    - [ ] Update CDK Exec Policy to NotebookEnabled=true
    - [ ]  Permissions Added to CDK Exec Policy
    - [ ]  Create Notebook Successfully
    - [ ] Delete Notebook Successfully
- Pipelines Test 
     - [ ]  Update CDK Exec Policy to PipelinesEnabled=true
     - [ ]  Permissions Added to CDK Exec Policy
    - [ ]   Create Pipeline Successfully
    - [ ]   Delete Pipeline Successfully
- Dashboards Test 
    - [ ]   Update CDK Exec Policy to DashboardsEnabled=true
    - [ ]   Permissions Added to CDK Exec Policy

File: deploy/cdk_exec_policy/cdkExecPolicy.yaml [hunk 5]
```diff
@@ -218,55 +288,68 @@ Resources:
...
-              - 'ec2:Modify*'
-              - 'ec2:Get*'
-              - 'ec2:*Address'
-              - 'elasticfilesystem:CreateFileSystem'
-              - 'elasticfilesystem:TagResource'
-              - 'elasticfilesystem:UntagResource'
-              - 'elasticfilesystem:DeleteFileSystem'
-            Resource: '*'
+          - !If 
+            - MLStudioEnabledCondition
```
Comment (noah-paige) [inline deploy/cdk_exec_policy/cdkExecPolicy.yaml:292]: Found some errors when creating Notebooks needing some additional EC2 permissions

The quick fix is to switch `MLStudioEnabledCondition` with `SagemakerComponentsEnabledCondition` (tested and all works then)

We could theoretically also determine the precise permissions needed for each Notebook and MLStudio (MLStudio permissions likely being a superset of the notebook permissions plus others) but it depends on the level of granularity we are wanting here

I am good to just pass `SagemakerComponentsEnabledCondition` unless others disagree

**Review by noah-paige — APPROVED** [review]
Tested and all looks good - approving
